File: Codes/jxdeng3264/000-100/001.Two-Sum/Solution.cpp
//ID:001.Two Sum
//Author:jxdeng3264@163.com
//Alg&Ds:Binary Search
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> t = nums;
        sort(nums.begin(), nums.end());

        vector<int> res;
        for (int i=0; i<nums.size()-1; ++i)
        {
            int sp = i+1;
            int ep = nums.size()-1;
            while (sp<=ep)
            {
                int mid =(sp+ep)/2;
                if (nums[mid]+nums[i]==target)
                {
                    res.push_back(nums[mid]);
                    res.push_back(nums[i]);
                    break;
                }

                if (nums[mid]<target-nums[i])
                    sp = mid+1;
                else
                    ep = mid-1;
            }
        }

        for (int i=0; i<t.size(); ++i)
            if (t[i]==res[0])
            {
                res[0] = i;
                break;
            }

        for (int i=0; i<t.size(); ++i)
            if (t[i]==res[1] && i!=res[0])
            {
                res[1] = i;
                break;
            }
        return res;
    }
};

File: Codes/jxdeng3264/000-100/001.Two-Sum/Solution.py
#!/usr/bin/env python3
# encoding: utf-8
# author: pyclearl

"""
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
"""


class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype List[int]

        >>> s = Solution()
        >>> nums = [2, 7, 11, 15]
        >>> target = 9
        >>> s.twoSum(nums, target)
        [0, 1]

        Because nums[0] + nums[1] = 2 + 7 = 9,
        """
        for i, n in enumerate(nums):
            try:
                # 求n的补在nums中是否存在, 存在就返回，可能越界
                index = nums.index(target - n)
                if index != i:
                    return [i, index]
            except ValueError as e:
                pass


if __name__ == '__main__':
    import doctest
    doctest.testmod(verbose=True)
